src/cparsec3/stream/stream_string.c

```diff
@@ -106,23 +63,32 @@ static Maybe(Tuple(Token(String), String)) take1(String s) {
  *            - nothing (if `n > 0` and `s` was empty)
  */
 static Maybe(Tuple(Tokens(String), String)) takeN(int n, String s) {
+  assert(s && "Null pointer");
   typedef Maybe(Tuple(Tokens(String), String)) R;
   if (n <= 0) {
-    Tokens(String) x = {.length = 0, .value = ""};
-    return (R){.value = {{x}, {s}}};
+    return (R){.value = {{""}, {s}}};
   }
-  if (empty(s)) {
+  if (null(s)) {
     return (R){.none = true};
   } else {
     int m = strnlen(s, n);
-    Tokens(String) x = {.length = m, .value = s};
-    return (R){.value = {{x}, {s + m}}};
+    char* chk = mem_malloc(m + 1);
+    assert(chk && "no memory");
+    strncpy(chk, s, m);
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Easily used incorrectly; doesn't always \0-terminate or check for invalid pointers [MS-banned] (CWE-120).  ](https://app.codacy.com/manual/mori0091/cparsec3/pullRequest?prid=4851446)